base/broadcast.jl

```diff
@@ -1211,4 +1211,29 @@ end
 end
 @inline broadcasted(::S, f, args...) where S<:BroadcastStyle = Broadcasted{S}(f, args)
 
+function _lazy end  # only used in `@:` macro; does not have to be callable
+# wrap the Broadcasted object in a tuple to avoid materializing
+@inline broadcasted(::typeof(_lazy), x) = (x,)
+
+"""
+    @: broadcasting_expression
+
+Construct a non-materialized broadcasted object from a `broadcasting_expression`
+and [`instantiate`](@ref) it.
```

Comment:
I see this as being akin to a generator; that is, `generator : comprehension :: Broadcasted : broadcast`.  Perhaps that could give us a bit of a better language around this?  Maybe call it a broadcast generator?

Reply:
I agree that "non-materialized broadcasted object" is a bit too mouthful. It's nice to have a concise word for it.

> broadcast generator

`Broadcasted` is kind of like generators but I'm seeing the generators as a way to define iterators (i.e., [limited forms of the coroutines](https://en.wikipedia.org/wiki/Coroutine#Comparison_with_generators)).  From this point of view, I find it somewhat confusing.  But then `Base.Generator` is just a lazy version of `map` in Julia and it's pretty close to `Broadcasted`.  So maybe it's fine?  I slightly prefer "lazy broadcast" although the term "lazy" is too overloaded.

> `generator : comprehension :: Broadcasted : broadcast`

By the way, what does `:` and `::` mean here?  Maybe `Base.Generator` corresponds to `Broadcasted` and `[f(x) for x in xs]` corresponds to `broadcast(f, xs)`?

Reply:
Those colons aren't code... it's a (now antiquated) way of writing a chained analogy. For example _puppy : dog :: kitten : cat_ can be read "a puppy is to a dog as a kitten is to a cat."  Or in other words, the pair of generator/comprehension are comparable to the pair Broadcasted/broadcast.  I shouldn't have written it like that — someone would only have reason to know that if they specifically took a particular US college entrance exam before 2005. 😄

Reply:
Ah, that's an interesting notation. Thanks for using it :smile: